### PR TITLE
Update cses 1625

### PR DESCRIPTION
**Pull request by vivianhan404**

### CODE REVIEW

**Comment by ryanchou-dev** [comment]
did you not update the module?

File: solutions/cses-1625.mdx [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+			for (int i = 0; i < 9; i++) {
+				onPath[0][i] = true;
+				onPath[8][i] = true;
+				onPath[i][0] = true;
+				onPath[i][8] = true;
+			}
+
+//            long mid = System.currentTimeMillis();
+//            System.out.println("add borders: " + (mid-start));
+
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:251]: uuuhhh

File: solutions/cses-1625.mdx [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	static boolean[][] onPath = new boolean[9][9];
+	static int[] dr = {-1, 0, 1, 0};
+	static int[] dc = {0, 1, 0, -1};
+	static int[] p = new int[48];
+
+	public static void main(String[] args) {
+		try {
+//      this is to figure out how long it takes for the program to run
+//      long start = System.currentTimeMillis();
+
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:226]: remove the try catch?

File: solutions/cses-1625.mdx [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.util.Scanner;
+
+public class GridPaths4 {
+
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:216]:
```suggestion
public class GridPaths {
```

File: solutions/cses-1625.mdx [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+		//turn already determined:
+		if (p[pathIdx] < 4) {
+			int nxtR = curR + dr[p[pathIdx]];
+			int nxtC = curC + dc[p[pathIdx]];
+			if (!onPath[nxtR][nxtC]) {
+				ret += tryPath(pathIdx + 1, nxtR, nxtC);
+			}
+		}
+		else if ((curC > 2) && onPath[curR][curC - 2]
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:287]: i
i really don't like this brace style
```java
if () {
} else if () {
} else {
}
```

File: solutions/cses-1625.mdx [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+<CPPSection>
+
+```cpp
+#include <iostream>
+using namespace std;
+
+int dr[4] = {-1, 0, 1, 0};
+int dc[4] = {0, 1, 0, -1};
+int p[48];
+bool onPath[9][9];
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:101]: magic numbers bad, make constants please

File: solutions/cses-1625.mdx [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+int tryPath(int pathIdx, int curR, int curC);
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:102]: maybe just put the function on top idk

File: solutions/cses-1625.mdx [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+		char cur = line[i];
+
+		if (cur == 'U') p[i] = 0;
+		else if (cur == 'R') p[i] = 1;
+		else if (cur == 'D') p[i] = 2;
+		else if (cur == 'L') p[i] = 3;
+		else p[i] = 4;
+	}
+
+	//set borders of grid
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:120]: put a space after the `//` before the comment

File: solutions/cses-1625.mdx [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+			onPath[i][j] = false;
+		}
+	}
+
+	int startIdx = 0;
+	int startR = 1;
+	int startC = 1;
+	int ans = tryPath(startIdx, startR, startC);
+	cout << ans << endl;
+	return 0;
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:138]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cses-1625.mdx [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+• running time: 0.6 seconds
+
+• number of recursive calls: 69 million
+
+Now is a good moment to stop optimizing the algorithm and see what we have
+achieved. The running time of the original algorithm was 483 seconds, and now
+after the optimizations, the running time is only 0.6 seconds. Thus, the
+algorithm became nearly 1000 times faster after the optimizations.
+
+<Info title="">
```
Comment (maggie-j-liu) [inline solutions/cses-1625.mdx:66]:
```suggestion
<Info>
```

**Review by maggie-j-liu — COMMENTED** [review]
i think the images uploaded are for a different problem?

**Comment by thecodingwizard** [comment]
@SansPapyrus683

File: solutions/cses-1625.mdx [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+	onPath[curR][curC] = true;
+
+	// turn already determined:
+	if (p[pathIdx] < 4) {
+		int nxtR = curR + dr[p[pathIdx]];
+		int nxtC = curC + dc[p[pathIdx]];
+		if (!onPath[nxtR][nxtC]) {
+			ret += tryPath(pathIdx + 1, nxtR, nxtC);
+		}
+	} /* // this next bit is not necessary for C++, but I left it in for reference
```
Comment (ryanchou-dev) [inline solutions/cses-1625.mdx:134]:
```suggestion
	} // this next bit is not necessary for C++, but I left it in for reference
```

File: solutions/cses-1625.mdx [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+	else if ((curC > 2) && onPath[curR][curC - 2]
+			&& (onPath[curR - 1][curC - 1] || onPath[curR + 1][curC - 1])
+			&& (!onPath[curR][curC - 1]))
+	{ // potential dead end on the left:
+		int nxtR = curR;
+		int nxtC = curC - 1;
+		ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	}
+	else if ((curC < 6) && onPath[curR][curC + 2]
+			&& (onPath[curR - 1][curC + 1] || onPath[curR + 1][curC + 1] )
```
Comment (ryanchou-dev) [inline solutions/cses-1625.mdx:144]:
```suggestion
			&& (onPath[curR - 1][curC + 1] || onPath[curR + 1][curC + 1])
```

**Review by ryanchou-dev — APPROVED** [review]
why is no one reviewing this? ❤️ the editorial :D

File: solutions/cses-1625.mdx (hunk 10 above)
Comment (ryanchou-dev) [reply solutions/cses-1625.mdx:134]: oops you can disregard this

**Comment by SansPapyrus683** [comment]
uuuuhhh i'll review in like an hour

**Comment by thecodingwizard** [comment]
@cpinitiative/content-managers this PR seems more or less fine except for a few small issues (I think?); can we just make any necessary changes for them? Feel free to delegate

File: solutions/cses-1625.mdx [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+		int nxtC = curC + 1;
+		ret += tryPath(pathId x +1, nxtR, nxtC);
+	}
+	else if ((curR > 2) && onPath[curR - 2][curC] && onPath[curR - 1][curC - 1]
+			&& (!onPath[curR - 1][curC]))
+	{ // potential dead end upwards
+		int nxtR = curR - 1;
+		int nxtC = curC;
+		ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	} */
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:157]: can you just remove this entire multiline

File: solutions/cses-1625.mdx [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+			ret += tryPath(pathIdx + 1, nxtR, nxtC);
+		}
+	}
+	// reset and return
+	onPath[curR][curC] = false;
+	return ret;
+}
+
+int main()
+{
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:172]: again, be consistent- either put all braces on a newline or all braces on the same line

File: solutions/cses-1625.mdx [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+	string line;
+	getline(cin, line);
+
+	// convert path to ints
+	for (int i = 0; i < 48; i++) {
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:177]: uh what does 48 mean

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cses-1625.mdx (hunk 14 above)
Comment (vivianhan404) [reply solutions/cses-1625.mdx:177]: The path is 48 characters long. It's not an input variable given by the problem because ALL possible inputs for this problem are 48 characters long.

File: solutions/cses-1625.mdx [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+	int ret = 0;
+	onPath[curR][curC] = true;
+
+	// turn already determined:
+	if (p[pathIdx] < 4) {
+		int nxtR = curR + dr[p[pathIdx]];
+		int nxtC = curC + dc[p[pathIdx]];
+		if (!onPath[nxtR][nxtC]) ret += tryPath(pathIdx + 1, nxtR, nxtC);
+	} 
+  // see Java solution for optimization 4 implementation
```
Comment (SansPapyrus683) [inline solutions/cses-1625.mdx:128]: indentation seems hecked up

File: solutions/cses-1625.mdx (hunk 14 above)
Comment (SansPapyrus683) [reply solutions/cses-1625.mdx:147]: can you add a comment explaining that

File: solutions/cses-1625.mdx (hunk 15 above)
Comment (vivianhan404) [reply solutions/cses-1625.mdx:128]: 1. Thank you for not swearing.
2. Is it better now? I couldn't tell which lines in specific you were pointing to.

Comment (maggie-j-liu) [reply solutions/cses-1625.mdx:128]: no, the indentation is still wrong on line 128

Comment (vivianhan404) [reply solutions/cses-1625.mdx]: Indentation issues for all comments _should_ be fixed now?

File: solutions/cses-1625.mdx [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+	}
+
+	public static int tryPath(int pathIdx, int curR, int curC) {
+		if ((onPath[curR][curC - 1] && onPath[curR][curC + 1])
+			&& (!onPath[curR - 1][curC] && !onPath[curR+1][curC])) return 0;
+		if ((onPath[curR - 1][curC] && onPath[curR + 1][curC])
+			&& (!onPath[curR][curC - 1] && !onPath[curR][curC + 1])) return 0;
+
+		if (curR == 7 && curC == 1) { // reached endpoint
+			if (pathIdx == p.length) return 1; //visited every cell -> valid!
```
Comment (dongliu0426) [inline solutions/cses-1625.mdx:227]: be consistent on adding a space after `//`

File: solutions/cses-1625.mdx [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+The consequences of this discrepancy which will be addressed later in the explanation.
+</Warning>
+
+### Basic Algorithm
+
+The first version of the algorithm does not contain any optimizations. We simply
+use backtracking to generate all possible paths from the upper-left corner to
+the lower-right corner and count the number of such paths.
+
+• running time: 483 seconds
```
Comment (dongliu0426) [inline solutions/cses-1625.mdx:25]: use `-` or `*`? I don't think `•` renders as a list.

Comment (dongliu0426) [reply solutions/cses-1625.mdx:25]: (same for below)

File: solutions/cses-1625.mdx [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+For C++ users, this will be enough pruning to make the program run in time under
+worst case conditions. However, for Java and Python users, this is still too
+slow (Java takes around 1.2 seconds for the worst case). Thus, we will need to
+further optimize our search.
+
+### Optimization 4
+
+If the path creates a dead end that is not the bottom left corner, either the
+path will fail to visit all squares (the path may stop at the dead end or pass
+over it, sealing a square off) or the path will end in the wrong location. Thus
```
Comment (dongliu0426) [inline solutions/cses-1625.mdx:82]:
```suggestion
over it, sealing a square off) or the path will end in the wrong location. Thus,
```

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: solutions/cses-1625.mdx [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+			int nxtR = curR;
+			int nxtC = curC + 1;
+			ret += tryPath(pathIdx + 1, nxtR, nxtC);
+		}
+		else if ((curR > 2) && onPath[curR - 2][curC] && onPath[curR - 1][curC - 1]
+			&& (!onPath[curR - 1][curC])) {  // potential dead end upwards
+		// note: I didn't include all possible scenarios because
+		// it wasn't necessary in order for the program to run in time
+			int nxtR = curR - 1;
+			int nxtC = curC;
```
Comment (ryanchou-dev) [inline solutions/cses-1625.mdx:259]: maybe indent the comment so it lines up with the code?

**Review by ryanchou-dev — APPROVED** [review]
❤️

**Comment by ryanchou-dev** [comment]
what's going on with the build?

File: solutions/cses-1625.mdx [hunk 20]
```diff
@@ -0,0 +1,281 @@
...
+- number of recursive calls: 221 million
+
+### Optimization 3
+
+The idea of Optimization 3 can be generalized: if the path cannot continue
+forward but can turn either left or right, the grid splits into two parts that
+both contain unvisited squares. It is clear that we cannot visit all squares
+anymore, so we can terminate the search.
+
+- running time: 0.6 seconds
```
Comment (nwatx) [inline solutions/cses-1625.mdx:54]: would it make sense to capitalize the first word for all lists?

also it would be nice but not required to texify the #

Comment (nwatx) [reply solutions/cses-1625.mdx:54]: e.g 0.6 in this case

File: solutions/cses-1625.mdx [hunk 21]
```diff
@@ -0,0 +1,281 @@
...
+- number of recursive calls: 69 million
+
+Now is a good moment to stop optimizing the algorithm and see what we have
+achieved. The running time of the original algorithm was 483 seconds, and now
+after the optimizations, the running time is only 0.6 seconds. Thus, the
+algorithm became nearly 1000 times faster after the optimizations.
+
+<Info>
+This is a usual phenomenon in backtracking, because the search tree is usually
```
Comment (nwatx) [inline solutions/cses-1625.mdx:63]: first part of the sentence is slightly unnecessary, but up to you to remove it.

File: solutions/cses-1625.mdx [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.util.Scanner;
+
+public class GridPaths {
+
+	static boolean[][] onPath = new boolean[9][9];  // added border to all four sides so a 7x7 becomes a 9x9
```
Comment (nwatx) [inline solutions/cses-1625.mdx:184]: > 80 chars

can we shorten this to less than 80 character line length?

File: solutions/cses-1625.mdx [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+
+		// convert path to ints
+		for (int i = 0; i < p.length; i++) {
+			char cur = line.charAt(i);
+
+			if (cur == 'U') p[i] = 0;
+			else if (cur == 'R') p[i] = 1;
+			else if (cur == 'D') p[i] = 2;
+			else if (cur == 'L') p[i] = 3;
+			else p[i] = 4; //cur == '?'
```
Comment (nwatx) [inline solutions/cses-1625.mdx:201]:
```suggestion
			else p[i] = 4; // cur == '?'
```

**Review by nwatx — CHANGES_REQUESTED** [review]

**Comment by vivianhan404** [comment]
Why is the build failing ;-; help pls

**Comment by maggie-j-liu** [comment]
hmm, not sure why the build is failing or what the error means. can you try merging master into your branch to see if it will fix it?

**Review by nwatx — APPROVED** [review]

**Comment by nwatx** [comment]
^ if build is fixed

**Review by dutinmeow — APPROVED** [review]
good luck with the build

**Review by maggie-j-liu — APPROVED** [review]
